### ISSUE TITLE
Use of `isize` for count changes should be generic

### ISSUE THREAD

**Issue by frankmcsherry** [frankmcsherry/differential-dataflow#16]
Currently differential dataflow uses `isize` to represent changes to counts. This allows it to handle a broad range of changes, but it is limiting in a few ways and could fairly easily become generic.

1. For many applications, an `i32` is sufficient. In fact, for many applications an `i8` is sufficient. Given the pervasive role of these differences, the size they take can matter. Several graph processing examples would prefer something like `i32` for their edge collections, as each edge does not change its frequency so much. Even an `i8` would be useful if it was stored separately from edge data (otherwise the `(edge, diff)` pair would remain the same size).

2. Several applications can make use of more sophisticated counts. For example, to compute a sum or average instead of `isize` one could use `(isize, isize)` where one field is the count and the other field is the sum (or change to it). Updates with the same key should be able to compact using the natural `+=` interpretation on pairs (or tuples, whatever) of numbers that can be added.

To be interpreted as a "diff", it seems (at first) that we just need a trait like so 

```rust
pub trait Diff : Add<Self, Output=Self> + Neg<Output=Self> + Zero;
```

The `Add` and `Neg` constraints are there so that we can manipulate the updates, and `Zero` is there so that we can tell when it is safe to drop the update on the floor. `Zero` is unstable (`Zeroable` would be great to have, for the `Option` optimization, but `NonZero` is unstable too). Perhaps we could use `Default` in the interim, or we could add an `is_zero` method to the trait, as the implementations will be mostly trivial.

The `join` implementation currently multiplies differences, and I'm not exactly certain what the intended semantics are here. If you join two collections without first calling `count` you get the product of the collections with the counts multiplied; should you expect something different if you join two collections without first calling `sum`? 

Perhaps a related question is "would these weird counts be useful anywhere other than leading in to an aggregation like `sum`?" I can't think of anything at the moment, but that may not be a great reason to hide them.

**Edit**: moving away from `isize` does raise the question of under- and over-flow. If we use `i8`s for diffs and an operator needs more, what do we do? One natural choice is to produce as many diffs as are necessary to accumulate up to the change, but this does mean that various operators may need to become aware that there may be multiple entries for a given `(key, val, time, _)` pattern.

**Edit**: Working hypothesis: the trait we want should probably be called `Ring`, following https://en.wikipedia.org/wiki/Ring_(mathematics).

**Comment by frankmcsherry** [comment]
This would fix #6 somewhat cleanly, I think. In the problem example, data are sourced as pairs of type `(K,SaneF32)`, with the goal of accumulating the float components. Now, unfortunately floats do not form a ring (due to associativity, making different sums of the same numbers not cancel). If `SaneF32` is replaced by a fixed precision `i32` (or `i64` if needed) then we could ingest the data as

```rust
let (input, stream) = builder.new_input::<(K, (i32, isize))>();
let collections = Collection::new(stream);
```

The use of the collection would need to be one that accumulates in place (I'm not certain `arrange` does this at the moment, but it could with a few lines of code), at which point all instances of a single key would have both `i32` and `isize` components accumulated.

**Comment by frankmcsherry** [comment]
Fixed by #18.
